falcon/cluster/cluster.py

```diff
@@ -145,93 +187,138 @@ def _build_query_ann_index(
     """
-    filenames, identifiers, precursor_mzs, rts = [], [], [], []
+    # Read the spectra for the m/z bucket.
+    filenames, identifiers, bucket_ids, precursor_mzs, rts = [], [], [], [], []
+    spectra = nb.typed.List()
     indptr_i = 0
-    # Find neighbors per specified precursor m/z interval.
-    for pkl_filename in tqdm.tqdm(bucket_filenames, desc='Buckets queried',
-                                  unit='bucket'):
-        # Read the spectra for the m/z split.
-        spectra_split, precursor_mzs_split, rts_split = nb.typed.List(), [], []
-        with open(pkl_filename, 'rb') as f_in:
-            for spec_raw in pickle.load(f_in):
-                spec_processed = process_spectrum(spec_raw)
-                # Discard low-quality spectra.
-                if spec_processed is not None:
-                    spectra_split.append(spec_processed)
-                    filenames.append(spec_processed.filename)
-                    identifiers.append(spec_processed.identifier)
-                    precursor_mzs_split.append(spec_processed.precursor_mz)
-                    rts_split.append(spec_processed.retention_time)
-        if len(spectra_split) == 0:
-            continue
-        precursor_mzs.append(np.asarray(precursor_mzs_split))
-        rts.append(np.asarray(rts_split))
-        # Convert the spectra to vectors.
-        vectors_split = vectorize(spectra=spectra_split)
-        n_split, dim = vectors_split.shape
-        # Figure out a decent value for the n_list hyperparameter based on
-        # the number of vectors.
-        # Rules of thumb from the Faiss wiki:
-        # https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index#how-big-is-the-dataset
-        if n_split == 0:
-            continue
-        if n_split < 10**2:
-            # Use a brute-force index instead of an ANN index when there
-            # are only a few items.
-            n_list = -1
-        elif n_split < 10**6:
-            n_list = 2**math.floor(math.log2(n_split / 39))
-        elif n_split < 10**7:
-            n_list = 2**16
-        elif n_split < 10**8:
-            n_list = 2**18
-        else:
-            n_list = 2**20
-            if n_split > 10**9:
-                logger.warning('More than 1B vectors to be indexed, consider '
-                               'decreasing the ANN size')
-        # Create an ANN index using the inner product (proxy for cosine
-        # distance) for fast NN queries.
-        if n_list <= 0:
-            index = faiss.IndexIDMap(faiss.IndexFlatIP(dim))
-        else:
-            index = faiss.IndexIVFFlat(faiss.IndexFlatIP(dim), dim, n_list,
-                                       faiss.METRIC_INNER_PRODUCT)
-            index.nprobe = min(math.ceil(index.nlist / 8), n_probe)
-        # Compute cluster centroids.
+    msms_spectra = spectra_df.apply(spectrum.df_row_to_spec, axis=1).tolist()
+    for spec_raw in msms_spectra:
```

Comment:
**question:** This is now no longer batched I think? This might lead to excessive memory consumption when processing large datasets (i.e. 100s to 1000s of runs), so we need to think about how to do this.

Reply:
Currently, all spectra with the same charge are read from lance and processed before creating the index. Another approach is to read the spectra in batches and add them to the index. However, all data to train the index should be in memory. In the case of large datasets, we could train on a (random) subset instead?

Reply:
Yes, I feel that we should be a bit smart about it for (very) large datasets, where we don't need to have all spectra. A random subset in that case is indeed sufficient. The maximum number of spectra can be based on how much memory a single spectrum vector consumes and the Faiss guidelines.

Reply:
- Spectra are processed before adding them to the lance dataset. 
- Removed the training on a random subset again because of non-determinism, so all spectra are loaded in memory. (still need to figure out lance sampling)

Reply:
Fixed seed for lance sampling. Currently training the index on a subset (`50 * nlist` as recommended) of the data again. However, this influences the clustering performance.